Initial implementation of app plugin.

rtibbles opened the pull request:
### Summary
Initial implementation of app plugin for Android, MacOS etc.to toggle on app specific behaviour
Add `/app/api/initialize/<token>` endpoint that allows for initialization of the app view and setting the cookie.
Has an optional `next` GET param that will redirect to a specific location.

### Reviewer guidance
Is having a registry the right way to conditionalize the behaviour depending on the OS?

Get the token for the URL from Kolibri shell:
```
from kolibri.core.device.models import DeviceAppKey
DeviceAppKey.get_app_key()
```
Put it in the above endpoint - you should get a cookie set `app_key_cookie`.

Try again with `?next=/learn` appended, you should be redirected to learn.

### Contributor Checklist


PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

rtibbles committed:
Initial implementation of app plugin.

rtibbles committed:
Add device app key and endpoint for validating the app key.

rtibbles committed:
Add AppView permission to AppCommands.

rtibbles committed:
Check appview cookie in JS appview check.

rtibbles committed:
Make capability reading robust for tests.

rtibbles committed:
Lint migration.

jamalex commented on kolibri/core/assets/src/utils/appCapabilities.js:
```diff
@@ -0,0 +1,7 @@
+
+export default {
+  launchIntent: appCapabilities.launch_intent,
+};
```
I've been juggling in my head the need to define the set of supported capabilities centrally in `core` vs in a more P2P fashion between plugins (pub/sub sort of model). Was there a strong reason to prefer this direction?

Similarly, flagging to myself to check elsewhere to see how a plugin would handle the case of it wanting to use a capability that's not even defined in core (gracefully) -- with an eye to supporting external/public plugins down the road.

rtibbles replied:
Mostly because there wasn't a clear need for it to be pluggable, so it seemed to make it more complicated.

In this case because we are exporting an object, attempting to check an undefined capability would return a falsy value.

jamalex replied:
Ok, yeah, as long as it's got "graceful degradation" then that seems fine.

jamalex commented on kolibri/core/assets/src/utils/browser.js:
```diff
@@ -37,4 +38,10 @@ export function isEmbeddedWebView() {
 }
 
+export function isAppView() {
+  const isLocal = window.location.host !== '127.0.0.1' || window.location.host !== 'localhost';
```
Isn't this... a tautology? Do we need to remove the `!`'s?

jamalex replied:
Although I think I'd prefer not having this constraint anyway -- if the token is secure, this doesn't add security, and just makes remote debugging harder.

rtibbles replied:
Sure, can remove - I had initially added this before the shibboleth existence checking.

jamalex commented on kolibri/core/assets/src/utils/browser.js:
```diff
@@ -37,4 +38,10 @@ export function isEmbeddedWebView() {
 
+export function isAppView() {
+  const isLocal = window.location.host !== '127.0.0.1' || window.location.host !== 'localhost';
+  const hasWebViewCookie = Boolean(cookiejs.get('app_key_cookie'));
```
Rather than just the cookie being defined, I think checking that it's valid would be ideal.

Is that something we could add to the session endpoint? Have it include a boolean attribute like `secureAppContext`? Then we don't need to do anything at all with the cookie in the frontend (can just use the same function in the backend as we do in the permission check), and that boolean should be available to the frontend through the session data, right?

rtibbles replied:
Yeah, can add it to the session certainly - in theory could then also prevent the cookie being read on the client side so that it is less leakable.

jamalex commented on kolibri/core/assets/src/utils/browser.js:
```diff
@@ -37,4 +38,10 @@ export function isEmbeddedWebView() {
+export function isAppView() {
+  const isLocal = window.location.host !== '127.0.0.1' || window.location.host !== 'localhost';
+  const hasWebViewCookie = Boolean(cookiejs.get('app_key_cookie'));
+  return isEmbeddedWebView() && isLocal && hasWebViewCookie;
```
Similarly to the localhost note above, I think it'd probably be best not having this be contingent upon the `isEmbeddedWebView` directly, as that will make debugging things like the click-to-login stuff easier.

jamalex replied:
(there's a distinction between functionality that only cares about being in an app-level secure context, e.g. the click-to-login, and functionality that is only possible at all in the app context itself, e.g. the whatsapp sharing -- the latter can check both `secureAppContext` and `isAndroidWebView`)

jamalex replied:
(although even the latter wouldn't need to be hidden in the UI based on isAndroidWebView -- what really matters for that is whether the _server_ is running in an Android app service context with the intent-based sharing capability)

jamalex replied:
Seems to me like there are 3 distinct things, which can factor in differently for each feature:
1. Is it a secure app context?
2. Are we inside a webview?
3. Does the capability exist?

In production, we'd never have 1 without 2, so the main reason for allowing 1 without 2 is for development/debugging.

rtibbles replied:
Sure, can remove the extra condition!

jamalex commented on kolibri/plugins/app/api.py:
```diff
@@ -0,0 +1,77 @@
+    """
+
+    def has_permission(self, request, view):
+        return request.META.get("REMOTE_ADDR") == "127.0.0.1"
```
If we are going to keep a "local only" constraint, do we need to add "localhost" here too? Or does it always resolve the client address into an IP?

rtibbles replied:
It always resolves to the IP, yup.

jamalex replied:
If we're removing the check on the frontend, we should remove it here too, and just be shibboleth-based.

jamalex commented on kolibri/plugins/app/utils.py:
```diff
@@ -0,0 +1,42 @@
+    def launch_intent(self, filename, message):
+        if LAUNCH_INTENT not in self._capabilities:
+            raise NotImplementedError(
+                "Launch intents is not supported on this platform"
```
*is -> are
or
*Launch -> Launching

rtibbles replied:
or "Launch intents" :) will update

jamalex commented on kolibri/plugins/app/api.py:
```diff
@@ -0,0 +1,77 @@
+
+    permission_classes = (FromSameDevicePermission, FromAppViewPermission)
+
+    if LAUNCH_INTENT in interface:
```
Conditional method existence :exploding_head:

rtibbles replied:
Don't even expose the endpoint if you don't need to. SECURE!

jamalex replied:
![](https://media1.giphy.com/media/jUwpNzg9IcyrK/giphy.gif)

jamalex commented on kolibri/plugins/learn/assets/src/views/ContentPage.vue:
```diff
@@ -328,6 +327,20 @@
       },
+      launchIntent() {
+        return client.post({
+          path: urls['kolibri:kolibri.plugins.app:appcommands-launch_intent'](),
```
I feel like maybe these URLs should somehow "go through" (be generated by) something more centralized in the "capabilities" system, rather than being somewhat "hard-coded" here. Just trying to think about how to keep the cross-plugin interfaces as clean/robust as possible. Just flagging in case you might have ideas for that.

rtibbles replied:
Yeah, can make a helper function for this.

jamalex commented on kolibri/plugins/app/api.py:
```diff
@@ -0,0 +1,77 @@
+            allowed_hosts={request.get_host()},
+            require_https=request.is_secure(),
+        ):
+            redirect_url = request.META.get("HTTP_REFERER")
```
For our current purposes, don't think we need the HTTP_REFERER fallback, but I guess it doesn't hurt.

jamalex reviewed (changes requested):
A few suggestions -- can chat to brainstorm if useful!

jamalex commented on kolibri/plugins/learn/assets/src/views/ContentPage.vue:
```diff
@@ -141,7 +139,10 @@
+  import client from 'kolibri.client';
+  import urls from 'kolibri.urls';
+  import { isEmbeddedWebView, isAppView } from 'kolibri.utils.browser';
+  import { launchIntent } from 'kolibri.utils.appCapabilities';
```
To the note above about what happens if a capability isn't defined, I'm assuming this would error if it's not defined?

rtibbles replied:
It would just be 'undefined'.

rtibbles committed:
Address PR comments.

indirectlylit commented on kolibri/core/assets/src/utils/appCapabilities.js:
```diff
@@ -0,0 +1,16 @@
+const appCapabilities = plugin_data.appCapabilities || {};
+
+export const can = {
+  launchIntent: appCapabilities.launch_intent,
```
is this capability really for generic launching of intents?

I would have expected one of:

* `share_file`
* `launch_android_intent`

or something more specific

jamalex replied:
Yeah, there could be other sorts of intents at some point, so my vote would be for something like "share_file".

rtibbles replied:
Fine by me, I was following the name of the function in the app-wrapper, so let's update that there too.

jamalex replied:
I had like 7 different names in there for different functions trying different approaches to the sharing until I finally got one to work reliably. Will be migrating it into the app plugin anyway, and will give it consistent naming there.

indirectlylit commented on kolibri/core/assets/src/utils/appCapabilities.js:
```diff
@@ -0,0 +1,16 @@
+  launchIntent: appCapabilities.launch_intent,
+};
+
+export const actionUrls = {
+  get launchIntent() {
+    return urls[`kolibri:kolibri.plugins.app:appcommands-launch_intent`]
+      ? urls[`kolibri:kolibri.plugins.app:appcommands-launch_intent`]()
+      : null;
+  },
+};
```
This is fine for now, but I don't know that `actionUrls` is quite the interface I would have expected `appCapabilities` to expose. I might have expected more of a set of Vuex actions or just plain functions that get called. For example, maybe something like:



```js
export const capabilityHandlers = {
  shareFile(filename, message) {
    const url = urls[`kolibri:kolibri.plugins.app:appcommands-share_file`];
    if (!url) {
      logger.warn("not implemented };
      return;
    }
    return client.post({
      path: url,
      method: 'POST',
      params: { filename, message },
    });
};
```

jamalex replied:
Yeah, exposing functions that then handle things internally seems like a nice, clean interface (minimal surface area, for easy migration paths).

rtibbles replied:
Sure, fine by me. Will update.

jamalex commented on kolibri/core/assets/src/utils/browser.js:
```diff
@@ -39,9 +39,9 @@ export function isEmbeddedWebView() {
 }
 
 export function isAppView() {
```
I think renaming this to `isAppContext` as well, for consistency, would be good.

rtibbles replied:
Done

jamalex commented on kolibri/core/device/utils.py:
```diff
@@ -75,3 +77,22 @@ def provision_device(**kwargs):
+
+
+def valid_app_key(app_key):
+    from .models import DeviceAppKey
```
Does this file get imported somewhere pre-app-loading, or why do these imports need to be local?

rtibbles replied:
To prevent circular imports.

jamalex commented on kolibri/core/device/utils.py:
```diff
@@ -75,3 +77,22 @@ def provision_device(**kwargs):
+    from .models import DeviceAppKey
+
+    response.set_cookie(APP_KEY_COOKIE_NAME, DeviceAppKey.get_app_key())
+    return response
```
Based on the function name, might be cleaner if it did the operation to the response but didn't return the response.

rtibbles replied:
Updated.

jredrejo commented on kolibri/core/assets/src/core-app/apiSpec.js:
```diff
@@ -101,6 +101,7 @@ import CatchErrors from '../utils/CatchErrors';
 import UiIconButton from '../views/KeenUiIconButton.vue';
 import UiToolbar from '../views/KeenUiToolbar.vue';
 import shuffled from '../utils/shuffled';
+import appCapabilities from '../utils/appCapabilities';
```
appCapabilities is not exporting a `default` function or const, thus this is not working

rtibbles replied:
Ah yep, I changed that in response to review comments. Am reverting back to having a default based on review above.

rtibbles replied:
Sorry, this should now be fixed!

rtibbles committed:
Update based on further feedback.

rtibbles commented:
Have addressed all review comments and manually tested as far as is possible.

Have confirmed that:
* The appContext getter is returning `true` when you have previously visited the initialize endpoint with the  correct shibboleth
* The app still loads and functions correctly.

Cannot confirm that the `shareFile` API is working as intended, as it requires it to have been registered on the backend during runtime.

jamalex commented on kolibri/plugins/app/api.py:
```diff
@@ -32,25 +23,23 @@ def has_permission(self, request, view):
 class AppCommandsViewset(ViewSet):
 
-    permission_classes = (FromSameDevicePermission, FromAppViewPermission)
+    permission_classes = (FromAppViewPermission,)
```
Should we also make this `FromAppContextPermission`, for full consistency?

jamalex replied:
(or just `AppContextPermission`)

rtibbles replied:
Sure

rtibbles committed:
Update naming of permission class.
